guide/adopting-the-configuration-cache.md

```diff
@@ -0,0 +1,384 @@
+|    `--- :jar                                                 (org.gradle.api.tasks.bundling.Jar)
+|         +--- :classes                                        (org.gradle.api.DefaultTask)
+|         |    +--- :compileJava                               (org.gradle.api.tasks.compile.JavaCompile)
+|         |    `--- :processResources                          (org.gradle.language.jvm.tasks.ProcessResources)
+|         `--- :compileJava                                    (org.gradle.api.tasks.compile.JavaCompile)
+`--- :check                                                    (org.gradle.api.DefaultTask)
+   +--- :checkstyleMain                                      (org.gradle.api.plugins.quality.Checkstyle)
+   |    +--- ...
+   +--- :checkstyleTest                                      (org.gradle.api.plugins.quality.Checkstyle)
+   |    +--- ...
+   +--- :spotlessCheck                                       (org.gradle.api.DefaultTask)
+   |    `--- ...
+   +--- :test                                                (org.gradle.api.tasks.testing.Test)
+   |    `--- ...
+   ...
+```
+
+When you first run a gradle task, the task graph is serialized and stored on disk. Upon a subsequent run of the same task, If none of these inputs have changed, Gradle skips the configuration phase entirely, loads the task graph from disk, and goes straight to the execution phase.
+
+The configuration phase typically runs faster than execution since heavy work belongs in the latter. However, without the cache, configuration phase is rerun with every single Gradle run. If you run a unit test multiple times, configuration is repeated, reproducing the same task graph every time. When the task itself is light, configuration can take up a large fraction of the latency. Configuration caching solves this by storing and reusing configuration results between runs, eliminating redundant work and speeding up iteration cycles.
+
+
+
+## Finding Configuration Cache problems
+
+To find Configuration Cache problems:
+
+1. Run `./gradlew build --configuration-cache`
+2. Check the output for any problems. If none appear, great! Your build is compatible.
+3. If not, fix the problems.
+4. Run `./gradlew build --configuration-cache` again. In our experience, this reveals additional issues not seen in the first run.
+
+The problems typically fall into three categories:
+
+1. **external process started `/usr/bin/git --version`**
+2. **cannot serialize object of type `org.gradle.api.internal.project.DefaultProject`, a subtype of `org.gradle.api.Project`, as these are not supported with the configuration cache**
+3. **invocation of `Task.project` at execution time is unsupported**
+
+The first error occurs when you start external processes during the configuration phase. Generally, it is preferred to run external processes in tasks with properly declared inputs and outputs to avoid unnecessary work when the task is UP-TO-DATE. 
```

Comment:
This is true, but there's also times where you do need to run things at configuration to decide when/what to configure a la palantir-java-format.

Reply:
True, but maybe that's a little bit too much detail for this guide?